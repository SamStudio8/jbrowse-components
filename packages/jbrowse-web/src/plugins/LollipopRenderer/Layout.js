export class PrecomputedFloatingLayout {
  constructor({ pairs, totalHeight }) {
    this.layout = new Map(pairs)
    this.totalHeight = totalHeight
  }

  add(uniqueId) {
    if (!this.layout.has(uniqueId))
      throw new Error(`layout error, precomputed layout is missing ${uniqueId}`)
  }

  getLayout() {
    return this.layout
  }

  getTotalHeight() {
    return this.totalHeight
  }

  static fromJSON(json) {
    return new PrecomputedFloatingLayout(json)
  }
}

export class FloatingLayout {
  constructor({ width }) {
    if (!width) throw new Error('width required to make a new FloatingLayout')
    this.width = width
  }

  items = []

  add(uniqueId, anchorLocation, width, height, data) {
    this.items.push([uniqueId, anchorLocation, width, height, data])
  }

  /**
   * @returns {Map} of uniqueId => {x,y,anchorLocation,width,height,data}
   */
  getLayout() {
    return new Map(
      this.items.map(([uniqueId, anchorLocation, width, height, data]) => [
        uniqueId,
        { x: anchorLocation, y: 0, width, height, data },
      ]),
    )
  }

  getTotalHeight() {
    return 100
  }

  toJSON() {
    return { pairs: [...this.getLayout()], totalHeight: this.getTotalHeight() }
  }

  static fromJSON() {
    throw new Error('not supported')
  }
}
