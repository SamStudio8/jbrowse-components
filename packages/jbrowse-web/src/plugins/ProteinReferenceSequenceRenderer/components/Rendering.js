import React from 'react'
import { observer } from 'mobx-react'
import ReactPropTypes from 'prop-types'

import { PropTypes as CommonPropTypes } from '../../../mst-types'
import { readConfObject } from '../../../configuration'
import { objectFromEntries } from '../../../util'
import aminoAcids from '../aminoAcids'
import { featuresConsensusSequence, contrastingTextColor } from '../util'

function Sequence({
  getColorForLetter,
  getTitleForLetter,
  region,
  config,
  bpPerPx,
  horizontallyFlipped,
  sequence,
}) {
  let s = sequence.split('')
  if (horizontallyFlipped) s = s.reverse()
  const height = readConfObject(config, 'height')
  const letterWidth = 1 / bpPerPx
  return (
    <>
      {s.map((letter, iter) => {
        const left = iter / bpPerPx
        const fill = getColorForLetter(letter)
        return (
          <g
            /* eslint-disable-next-line */
            key={`${region.start}-${iter}`}
            transform={`translate(${left}, 0)`}
          >
            <rect
              x={0}
              y={0}
              width={letterWidth}
              height={height}
              fill={fill}
              title={getTitleForLetter(letter)}
            />
            <text
              textAnchor="middle"
              x={letterWidth / 2}
              y={height * 0.8}
              fill={contrastingTextColor(fill)}
            >
              {bpPerPx < 0.1 ? letter : ''}
            </text>
          </g>
        )
      })}
    </>
  )
}

Sequence.propTypes = {
  region: CommonPropTypes.Region.isRequired,
  bpPerPx: ReactPropTypes.number.isRequired,
  horizontallyFlipped: ReactPropTypes.bool,
}
Sequence.defaultProps = {
  features: new Map(),
  horizontallyFlipped: false,
}

/**
 * given an iterable of features, yield only those features with
 * the requested type
 *
 * @param {string} type
 * @param {Iterable[SimpleFeature]} features
 * @returns {Iterable} of features in the iterator that have the requested type
 */
function* getFeaturesOfType(type, features) {
  for (const feature of features) {
    if (feature.get('type') === type) yield feature
  }
}

const aaColors = objectFromEntries(
  aminoAcids.map(aaRecord => [aaRecord.letter, aaRecord.color]),
)
const aaNames = objectFromEntries(
  aminoAcids.map(aaRecord => [aaRecord.letter, aaRecord.fullName]),
)

function Rendering(props) {
  const { bpPerPx, config, region, features } = props
  const height = readConfObject(config, 'height')
  const widthPx = (region.end - region.start) / bpPerPx
  const proteinSequence = featuresConsensusSequence(
    region,
    getFeaturesOfType('protein', features.values()),
  )
  return (
    <svg height={height} width={widthPx} style={{ fontSize: height * 0.9 }}>
      {bpPerPx >= 1 ? (
        <text className="blur">Zoom in to see sequence</text>
      ) : (
        <Sequence
          {...props}
          sequence={proteinSequence}
          getColorForLetter={letter => aaColors[letter] || '#ffffff'}
          getTitleForLetter={letter => aaNames[letter]}
        />
      )}
    </svg>
  )
}
Rendering.propTypes = {
  config: CommonPropTypes.ConfigSchema.isRequired,
  bpPerPx: ReactPropTypes.number.isRequired,
  region: ReactPropTypes.shape({ end: ReactPropTypes.number.isRequired })
    .isRequired,
}

export default observer(Rendering)
