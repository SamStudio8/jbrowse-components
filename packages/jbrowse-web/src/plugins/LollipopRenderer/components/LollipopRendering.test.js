import React from 'react'
import TestRenderer from 'react-test-renderer'
import Rendering from './LollipopRendering'
import SimpleFeature from '../../../util/simpleFeature'
import ConfigSchema from '../configSchema'
import { PrecomputedFloatingLayout, FloatingLayout } from '../Layout';

// these tests do very little, let's try to expand them at some point
test('no features', () => {
  const renderer = TestRenderer.create(
    <Rendering
      width={500}
      height={500}
      region={{ assemblyName: 'toaster', refName: 'zonk', start: 0, end: 300 }}
      layout={new PrecomputedFloatingLayout({ pairs: [], totalHeight: 20 })}
      config={{}}
      bpPerPx={3}
    />,
  )
  const result = renderer.toJSON()

  expect(result).toMatchSnapshot()
})

test('one feature', () => {
  const renderer = TestRenderer.create(
    <Rendering
      width={500}
      height={500}
      region={{ assemblyName: 'toaster', refName: 'zonk', start: 0, end: 1000 }}
      layout={new FloatingLayout({ width: 100 })}
      features={
        new Map([['one', new SimpleFeature({ id: 'one', start: 1, end: 3 })]])
      }
      config={ConfigSchema.create({})}
      bpPerPx={3}
    />,
  )
  const result = renderer.toJSON()

  expect(result).toMatchSnapshot()
})
