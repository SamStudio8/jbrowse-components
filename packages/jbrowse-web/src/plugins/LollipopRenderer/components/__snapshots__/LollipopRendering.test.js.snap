// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`no features 1`] = `
<svg
  className="LollipopRendering"
  height={20}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
  onMouseUp={[Function]}
  style={
    Object {
      "position": "relative",
    }
  }
  width={100}
/>
`;

exports[`one feature 1`] = `
<svg
  className="LollipopRendering"
  height={30}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
  onMouseUp={[Function]}
  style={
    Object {
      "position": "relative",
    }
  }
  width={333.3333333333333}
>
  <circle
    cx={0.6666666666666661}
    cy={15}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
    r={15}
    style={
      Object {
        "fill": "green",
      }
    }
    title="one"
  />
</svg>
`;
